File: merge-two-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
# recursively is easiest...
# 1 2 4
#       t
#       b
# 1 3 4 5 6 7
# d - 1 1 2 3 4 4 5

# 1
#   a
# b
# 2 3

# 1- 
class Solution:
    def mergeTwoLists(self, a: Optional[ListNode], b: Optional[ListNode]) -> Optional[ListNode]:
        if not a and not b:
            return None
        if not a or not b:
            return a or b
        if a.val < b.val:
            a.next = self.mergeTwoLists(a.next, b)
            return a
        b.next = self.mergeTwoLists(a, b.next)
        return b
        

File: lc/go/merge-two-sorted-lists.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 1 2 4
   ^
 1 3 4
 ^

loop until a and b are both nil
    aval = bval = 101
    reset them if the node vals if not nil
    if aval < bval
        r.next = a
        a = a.next
    else 
        r.next = b
        b = b.next

 return r

 1
 2
 av bv = 1 2
 r = (0) -> 1
 */
func mergeTwoLists(a *ListNode, b *ListNode) *ListNode {
    r := &ListNode{}
    rr := r
    inf := math.MaxInt
    for ; a != nil || b != nil; r = r.Next {
        aval, bval := inf, inf
        if a != nil {aval = a.Val}
        if b != nil {bval = b.Val}
        if aval < bval {
            r.Next = a
            a = a.Next
        } else {
            r.Next = b
            b = b.Next
        }
    }
    return rr.Next
}
